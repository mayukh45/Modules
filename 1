##========================================================================================
##========================================================================================
## Start of BASIC ACCESS  Interface Bus specification 
#
##========================================================================================
##========================================================================================

astob:
          wr:
                 data:
                           direction: input
                           width : 18
                           type: fluid


                 valid:
                           direction: input
                           width : 1
                           type: nonfluid

                 ready:
                           direction: output
                           width : 1
                           type: nonfluid


          rd:
                data:
                           direction: output
                           width : 18
                           type: fluid

                valid:
                           direction: output
                           width : 1
                           type: nonfluid

                ready:
                           direction: input
                           width : 1
                           type: nonfluid

          snoop:
                data:
                           direction: input
                           width : 18
                           type: fluid

                valid:
                           direction: input
                           width : 1
                           type: nonfluid

                match:
                           direction: input
                           width : 1
                           type: nonfluid
