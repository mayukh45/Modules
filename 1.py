Arbiter.py:    def shift_request(self):
Arbiter.py:    def shift_grant(self):
Arbiter.py:    def grant_signal(self):
Arbiter.py:    def rotate(self):
Arbiter.py:    def declare_weight_inputs(self):
Arbiter.py:    def declare_weight_vectors(self):
Arbiter.py:    def declare_weight_registers(self):
Arbiter.py:    def get_body(self):
Arbiter.py:    def update_weight_code(self):
Arbiter.py:    def __str__(self):
CAM.py:    def get_body(self):
CAM.py:    def request_declaration(self):
CAM.py:    def cam_locations(self):
CAM.py:    def cam_write(self):
CAM.py:    def assign_snoop(self):
CAM.py:    def snoop_cam(self):
CAM.py:    def assign_freeup(self):
CAM.py:    def __str__(self):
Encoder.py:    def __init__(self):
Encoder.py:    def binary2gray(self):
Encoder.py:    def gray2binary(self):
Encoder.py:    def onehot2binary(self):
Encoder.py:    def get_body(self):
Encoder.py:    def __str__(self):
Multiplexor.py:    def get_body(self):
Multiplexor.py:    def declare_ports(self):
Multiplexor.py:    def declare_vectors(self):
Multiplexor.py:    def assign_data(self):
Multiplexor.py:    def __str__(self):
OrderingPoint.py:    def module_declaration(self):
OrderingPoint.py:    def io_declaration(self):
OrderingPoint.py:    def assign_wires(self):
OrderingPoint.py:    def get_body(self):
OrderingPoint.py:    def __str__(self):
PacketConverter.py:    def get_body(self):
PacketConverter.py:    def declaration(self):
PacketConverter.py:    def add_collated(self):
PacketConverter.py:    def assign_data(self):
PacketConverter.py:    def __str__(self):
Pipe.py:    def __init__(self):
Pipe.py:    def pipe(self):
Pipe.py:    def skid(self):
Pipe.py:    def get_body(self):
Pipe.py:    def __str__(self):
smartasic.py:    def get_declaration(self):
smartasic.py:    def __str__(self):
smartasic.py:    def get_port_str(self):
smartasic.py:    def get_header(self):
smartasic.py:    def update(self):
smartasic.py:    def get_body(self):
smartasic.py:    def get_fifo_v(self):
smartasic.py:    def __str__(self):
Snoopable_FIFO.py:    def get_body(self):
Snoopable_FIFO.py:    def declare_register(self):
Snoopable_FIFO.py:    def assign_loc(self):
Snoopable_FIFO.py:    def write_loc(self):
Snoopable_FIFO.py:    def assign_read(self):
Snoopable_FIFO.py:    def assign_snoop_match(self):
Snoopable_FIFO.py:    def __str__(self):
