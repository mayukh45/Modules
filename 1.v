
module AHLruArbitor_8 (clk
,rstn
,req);

input clk;
input rstn;
input [31:0] req;

module lru_arbiter (
        clk,
        rstn,
        req,
        gnt_busy,
        gnt
);


input clk, rstn
input [NUMCLIENTS - 1:0] req;
input [NUMCLIENTS - 1:0] gnt_busy;
output [NUMCLIENTS - 1:0] gnt;

reg [NUMCLIENTS - 1:0] req0_used_status;
reg [NUMCLIENTS - 1:0] req1_used_status;
reg [NUMCLIENTS - 1:0] req2_used_status;
reg [NUMCLIENTS - 1:0] req3_used_status;

reg [8:0] req0;
reg [8:0] req1;
reg [8:0] req2;
reg [8:0] req3;
reg [8:0] req4;
reg [8:0] req5;
reg [8:0] req6;
reg [8:0] req7;_used_status

always @(posedge clk, negedge rstn) begin
        if(~rstn) begin
        req0_used_status <= 4'd4;
        req1_used_status <= 4'd3;
        req2_used_status <= 4'd2;
        req3_used_status <= 4'd1;
	//STATUS_USED
        end
        else begin
	//GNT_STATUS
                if(gnt[0]& !gnt_busy) begin

                        req0_used_status <= 4'b1;
                        req1_used_status <= req1_used_status<req0_used_status ? req1_used_status + 1'b1 : req1_used_status;
                        req2_used_status <= req2_used_status<req0_used_status ? req2_used_status + 1'b1 : req2_used_status ;
                        req3_used_status <= req3_used_status<req0_used_status ? req3_used_status + 1'b1 : req3_used_status ;
                end
                if(gnt[1] & !gnt_busy) begin
                        req1_used_status <= 4'b1;
                        req0_used_status <= req0_used_status<req1_used_status ? req0_used_status + 1'b1 : req0_used_status;
                        req2_used_status <= req2_used_status<req1_used_status ? req2_used_status + 1'b1 : req2_used_status ;
                        req3_used_status <= req3_used_status<req1_used_status ? req3_used_status + 1'b1 : req3_used_status ;
                end
                if(gnt[2]& !gnt_busy) begin
                        req2_used_status <= 4'b1;
                        req0_used_status <= req0_used_status<req0_used_status ? req0_used_status + 1'b1 : req0_used_status;
                        req1_used_status <= req1_used_status<req0_used_status ? req1_used_status + 1'b1 : req1_used_status ;
                        req3_used_status <= req3_used_status<req0_used_status ? req3_used_status + 1'b1 : req3_used_status ;
                end
                if(gnt[3] & !gnt_busy) begin
                        req3_used_status <= 4'b1;
                        req0_used_status <= req0_used_status<req0_used_status ? req0_used_status + 1'b1 : req0_used_status;
                        req1_used_status <= req1_used_status<req1_used_status ? req1_used_status + 1'b1 : req1_used_status ;
                        req2_used_status <= req2_used_status<req2_used_status ? req2_used_status + 1'b1 : req2_used_status ;
                end

        end
end

always @(req, gnt_busy) begin
//GNT_REQUEST
        gnt_pre[3:0] = 4'd0;
        req_int [3:0]= req[3:0] & {4{gnt_busy}};
        if(req[0]) begin
                gnt_pre[0] = (req0_used_status==4) ? 1'b1 :
                             ((req1 & (req1_used_status > req0_used_status)) ? 1'b0 :
                             ((req2 & (req2_used_status > req0_used_status)) ? 1'b0 :
                             ((req3 & (req3_used_status > req0_used_status)) ? 1'b0 :
                             1'b1;
                 end
        if(req[1]) begin
                gnt_pre[1] = (req1_used_status==4) ? 1'b1 :
                             ((req0 & (req0_used_status > req1_used_status)) ? 1'b0 :
                             ((req2 & (req2_used_status > req1_used_status)) ? 1'b0 :
                             ((req3 & (req3_used_status > req1_used_status)) ? 1'b0 :
                             1'b1;
                 end

        if(req[2]) begin
                gnt_pre[2] = (req2_used_status==4) ? 1'b1 :
                             ((req1 & (req1_used_status > req2_used_status)) ? 1'b0 :
                             ((req0 & (req0_used_status > req2_used_status)) ? 1'b0 :
                             ((req3 & (req3_used_status > req2_used_status)) ? 1'b0 :
                             1'b1;
                 end
        if(req[3]) begin
                gnt_pre[3] = (req3_used_status==4) ? 1'b1 :
                             ((req0 & (req0_used_status > req3_used_status)) ? 1'b0 :
                             ((req2 & (req2_used_status > req3_used_status)) ? 1'b0 :
                             ((req1 & (req1_used_status > req3_used_status)) ? 1'b0 :
                             1'b1;
                 end
end

assign gnt[NUMCLIENTS - 1:0] = gnt_pre[NUMCLIENTS - 1:0];

endmodule

endmodule

