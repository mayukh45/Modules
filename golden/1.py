AH_2OrderedSwitch.py:    def main(self):
AH_2OrderedSwitch.py-        self.get_body()
AH_2OrderedSwitch.py-
AH_2OrderedSwitch.py-    def __init__(self, number_of_egress,ds_packet_size, ups_packet_size, ds_decodable_field_width, ups_response_decodable_field_width, snoopdepth):
AH_2OrderedSwitch.py-        self.name = "AH_"+self.__class__.__name__+"_"+str(number_of_egress)+"_"+str(ds_packet_size)+"_"+str(ups_packet_size)+"_"+str(ds_decodable_field_width)+"_"+str(ups_response_decodable_field_width)
AH_2OrderedSwitch.py-        BasicModule.__init__(self, self.name)
AH_2OrderedSwitch.py-        self.NumberOfEgress = number_of_egress
AH_2OrderedSwitch.py-        self.DsPktSize = ds_packet_size
AH_2OrderedSwitch.py-        self.SnoopDepth = snoopdepth
AH_2OrderedSwitch.py-        self.upPktSize = ups_packet_size
AH_2OrderedSwitch.py-        self.dsDecodableFieldWidth = ds_decodable_field_width
AH_2OrderedSwitch.py-        self.upResponseDecodableFieldWidth = ups_response_decodable_field_width
AH_2OrderedSwitch.py-        self.port_dict = {}
--
AH_CAM.py:    def main(self):
AH_CAM.py-        self.write_to_file(self.get_verilog())
AH_CAM.py-        return self.get_verilog()
AH_CAM.py-
AH_CAM.py-
AH_CAM.py-    def __init__(self, camdepth, camwidth, snoopwidth, path_of_yaml=None,bus_name=None):
AH_CAM.py-        self.CamDepth = camdepth
AH_CAM.py-        self.EncodedDepth = ceil(log2(int(camdepth)))
AH_CAM.py-        self.SnoopWidth = snoopwidth
AH_CAM.py-        self.CamWidth = camwidth
AH_CAM.py-        if path_of_yaml is None:
AH_CAM.py-            path_of_yaml = "../../../smartasic2/refbuses/astob_for_order_switch.yaml"
AH_CAM.py-            bus_name = "astob"
--
AH_Decoder.py:    def main(self):
AH_Decoder.py-        print("I am about to write a verilog file now")
AH_Decoder.py-        self.write_to_file(self.get_verilog())
AH_Decoder.py-        print (self.get_verilog())
AH_Decoder.py-        return self.get_verilog()
AH_Decoder.py-
AH_Decoder.py-    def __init__(self, portwidth, numclients, path_of_yaml=None, bus_name=None, apertures_yaml=None):
AH_Decoder.py-        if path_of_yaml is None:
AH_Decoder.py-             path_of_yaml = "../../../smartasic2/refbuses/decoder.yaml"
AH_Decoder.py-             bus_name = "decoder"
AH_Decoder.py-             apertures_yaml = "decode_apertures.yaml"
AH_Decoder.py-
AH_Decoder.py-        self.AperturesYaml = apertures_yaml
--
AH_DivPipelined.py:    def main(self):
AH_DivPipelined.py-        self.write_to_file(self.get_verilog())
AH_DivPipelined.py-        return self.get_verilog()
AH_DivPipelined.py-
AH_DivPipelined.py-    def __init__(self,bits,stages,path_of_yaml=None,bus_name=None):
AH_DivPipelined.py-        self.bits=bits
AH_DivPipelined.py-        self.stages=stages
AH_DivPipelined.py-        self.body=""
AH_DivPipelined.py-        self.name="AH_"+self.__class__.__name__+"_"+str(bits)+"_"+str(stages)
AH_DivPipelined.py-        BasicModule.__init__(self , self.name)
AH_DivPipelined.py-        self.variable_dict={}
AH_DivPipelined.py-        BusParser.__init__(self,self.load_dict(path_of_yaml),bus_name)
AH_DivPipelined.py-        self.Create_dic_of_variable()
--
AH_LruArbiter.py:    def main(self):
AH_LruArbiter.py-        self.write_to_file(self.get_verilog())
AH_LruArbiter.py-        return self.get_verilog()
AH_LruArbiter.py-
AH_LruArbiter.py-    def __init__(self,num_clients,path_of_yaml=None,bus_name=None):
AH_LruArbiter.py-        self.bus_name=bus_name
AH_LruArbiter.py-        self.Num_Clients=num_clients
AH_LruArbiter.py-        self.variable_dict={}
AH_LruArbiter.py-        self.name="AH_"+self.__class__.__name__+"_"+str(num_clients)
AH_LruArbiter.py-        BasicModule.__init__(self, self.name)
AH_LruArbiter.py-        BusParser.__init__(self,self.load_dict(path_of_yaml),bus_name)
AH_LruArbiter.py-        self.body=""
AH_LruArbiter.py-        self.Create_dic_of_variable()
--
AH_Multiplexor.py:    def main(self):
AH_Multiplexor.py-        print("I am about to write a verilog file now")
AH_Multiplexor.py-        self.write_to_file(self.get_verilog())
AH_Multiplexor.py-        print (self.get_verilog())
AH_Multiplexor.py-        return self.get_verilog()
AH_Multiplexor.py-
AH_Multiplexor.py-    def __init__(self, portwidth, numclients, isdemux, isbinary, path_of_yaml=None, bus_name=None):
AH_Multiplexor.py-        self.PortWidth = portwidth
AH_Multiplexor.py-        self.NumClients = numclients
AH_Multiplexor.py-        self.IsDemux = isdemux
AH_Multiplexor.py-        self.body = None
AH_Multiplexor.py-        self.IsBinary = isbinary
AH_Multiplexor.py-
--
AH_Mutiplexor.py:    def main(self):
AH_Mutiplexor.py-        self.write_to_file(self.get_verilog())
AH_Mutiplexor.py-        return self.get_verilog()
AH_Mutiplexor.py-
AH_Mutiplexor.py-
AH_Mutiplexor.py-    def __init__(self, portwidth, numclients, isdemux, isbinary):
AH_Mutiplexor.py-        self.PortWidth = portwidth
AH_Mutiplexor.py-        self.NumClients = numclients
AH_Mutiplexor.py-        self.isDemux = isdemux
AH_Mutiplexor.py-        self.isBinary = isbinary
AH_Mutiplexor.py-        self.body=""
AH_Mutiplexor.py-        self.variable_dict={}
AH_Mutiplexor.py-        self.Create_dic_of_variable()
--
AH_OrderedSwitch.py:    def main(self):
AH_OrderedSwitch.py-        self.get_body()
AH_OrderedSwitch.py-
AH_OrderedSwitch.py-    def __init__(self, number_of_egress,ds_packet_size, ups_packet_size, ds_decodable_field_width, ups_response_decodable_field_width, snoopdepth):
AH_OrderedSwitch.py-        self.name = "AH_"+self.__class__.__name__+"_"+str(number_of_egress)+"_"+str(ds_packet_size)+"_"+str(ups_packet_size)+"_"+str(ds_decodable_field_width)+"_"+str(ups_response_decodable_field_width)
AH_OrderedSwitch.py-        BasicModule.__init__(self, self.name)
AH_OrderedSwitch.py-        self.NumberOfEgress = number_of_egress
AH_OrderedSwitch.py-        self.dsPktSize = ds_packet_size
AH_OrderedSwitch.py-        self.SnoopDepth = snoopdepth
AH_OrderedSwitch.py-        self.upPktSize = ups_packet_size
AH_OrderedSwitch.py-        self.dsDecodableFieldWidth = ds_decodable_field_width
AH_OrderedSwitch.py-        self.upResponseDecodableFieldWidth = ups_response_decodable_field_width
AH_OrderedSwitch.py-        self.dict = {}
--
AH_PacketConverter.py:    def main(self):
AH_PacketConverter.py-        self.write_to_file(self.get_verilog())
AH_PacketConverter.py-        return self.get_verilog()
AH_PacketConverter.py-
AH_PacketConverter.py-
AH_PacketConverter.py-
AH_PacketConverter.py-    def __init__(self, width1, width2, isW2N,path_of_yaml=None,bus_name=None):
AH_PacketConverter.py-        self.WideWidth = max(width1, width2)
AH_PacketConverter.py-        self.NarrowWidth = min(width1, width2)
AH_PacketConverter.py-        self.isW2N = isW2N
AH_PacketConverter.py-        self.ratio = (self.WideWidth // self.NarrowWidth)
AH_PacketConverter.py-        self.VecWidth = int(log2(self.WideWidth/self.NarrowWidth))
AH_PacketConverter.py-        if self.isW2N == "True":
--
AH_RoundRobinArbiter.py:    def main(self):
AH_RoundRobinArbiter.py-        self.write_to_file(self.get_verilog())
AH_RoundRobinArbiter.py-        return self.get_verilog()
AH_RoundRobinArbiter.py-
AH_RoundRobinArbiter.py-
AH_RoundRobinArbiter.py-    def __init__(self,num_clients,path_of_yaml=None,bus_name=None,weight=None):
AH_RoundRobinArbiter.py-        self.bus_name=bus_name
AH_RoundRobinArbiter.py-        self.Num_Clients=num_clients
AH_RoundRobinArbiter.py-        self.weight = weight
AH_RoundRobinArbiter.py-        if weight==None:
AH_RoundRobinArbiter.py-            self.name="AH_"+self.__class__.__name__+"_"+str(num_clients)
AH_RoundRobinArbiter.py-        else:
AH_RoundRobinArbiter.py-            self.name="AH_"+self.__class__.__name__+"_"+str(num_clients)+"_"+str(weight)
--
AH_SnoopableFIFO.py:    def main(self):
AH_SnoopableFIFO.py-        self.write_to_file(self.get_verilog())
AH_SnoopableFIFO.py-        return self.get_verilog()
AH_SnoopableFIFO.py-
AH_SnoopableFIFO.py-    def __init__(self, fifowidth, fifodepth, snoopwidth, path_of_yaml = None, bus_name = None):
AH_SnoopableFIFO.py-
AH_SnoopableFIFO.py-        self.FifoWidth = fifowidth
AH_SnoopableFIFO.py-        self.FifoDepth = fifodepth
AH_SnoopableFIFO.py-        self.SnoopWidth = snoopwidth
AH_SnoopableFIFO.py-        if path_of_yaml is None:
AH_SnoopableFIFO.py-            path_of_yaml = "../../../smartasic2/refbuses/astob.yaml"
AH_SnoopableFIFO.py-            bus_name = "astob"
AH_SnoopableFIFO.py-
--
smartasic.py:    def main(self):
smartasic.py-        self.write_to_file(self.get_verilog())
smartasic.py-        return self.get_verilog()
smartasic.py-
smartasic.py-
smartasic.py-
