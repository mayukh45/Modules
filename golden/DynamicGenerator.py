#=============================================================================================
# Class Dynamic Generator - The chief functionality of this class to be able to take a piece
# of code as string and able to run it as any other python code and return the output string
# We achieve this today by writing into py file and getting another output file and write
# back to main file.
# TODO: In future, find out a better way to achieve this.
#=============================================================================================

import os
from datetime import datetime   #### import datatime
class DynamicGenerator:
    """
    body_data -> actual template means if we go for LruArbiter module, we are passing body template of LruArbiter module.
    dic -> dictonary of variable we are using in Module, it is used when we are creating temporary python file.
    """
    def __init__(self,dic,body_data):
        self.body_data=body_data
        self.dic=dic
        self.temp_py_file=""
        self.temp_txt_file=""

    def parse_body(self):
        main_body=""
        split_body=self.body_data.split("/f_f/")    # /f_f/ is an expression, based on this we are splitinng the body into a list.

        for b in range(len(split_body)):
            """
            if the index is odd then we process odd position code, dumped into pyhton code then process into text file.
            otherwise just concat as string handleing

            """
            if b%2==0:
                main_body+=split_body[b]
            else:
                main_body+=self.temporary_pyfile(self.dic,split_body[b])
        return main_body


    def temporary_pyfile(self,dic,body_data):
        time_current=datetime.now().strftime('%H_%M_%S')
        code=""
        for x,y in dic.items():
            code+=x+"="+str(y)+"\n" ## here we process variable to temp python
        code+="\n"+body_data+"\nprint(code)"
        self.temp_py_file="temp_"+time_current+".py"    # creting temporay pyhton based on time to handle mutithreading
        with open(self.temp_py_file,'w') as f:
            f.write(code)
            f.close()
        self.temp_txt_file="output_"+time_current+".txt"    # creating temporary text file from where we retrive output and processed to verilog file.
        cmd="python "+self.temp_py_file
        os.system(cmd+'>'+self.temp_txt_file) # running python file using os.system
        return self.temporary_txtfile()

    """
    method for retrive output from temporary text file, and copy into body
    """
    def temporary_txtfile(self):
        lines=""
        fp=open(self.temp_txt_file,'r')
        for p in fp:
            lines+=p
        return lines

    """
    after processing all operation removing all temporary file
    """
    def silentremove(self):
        try:
            # os.remove(self.temp_py_file)
            # os.remove(self.temp_txt_file)
            os.system("rm temp_*")
            os.system("rm output_*")
        except:
            raise Exception("file not found")
