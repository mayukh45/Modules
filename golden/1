import os							import os
from smartasic import BasicModule,Port				from smartasic import BasicModule,Port
from DynamicGenerator import DynamicGenerator			from DynamicGenerator import DynamicGenerator
from BusParser import BusParser					from BusParser import BusParser
from pathlib import Path					from pathlib import Path
from math import log2, ceil					from math import log2, ceil
import sys							import sys


class CAM(BasicModule):						class CAM(BasicModule):
   								   
    #========================================================	    #========================================================
    #  								    #  
    #========================================================	    #========================================================
    def add_ports_from_bus(self, filepath, bus_name):		    def add_ports_from_bus(self, filepath, bus_name):
        parser = BusParser(filepath, bus_name)			        parser = BusParser(filepath, bus_name)
        parser.add_sub_dict_flat("snoop" , {"sin" : {"directi	        parser.add_sub_dict_flat("snoop" , {"sin" : {"directi
        parser.wid_op_flat("wdata",self.CamWidth)		        parser.wid_op_flat("wdata",self.CamWidth)
        parser.wid_op_flat("sin",self.SnoopWidth)		        parser.wid_op_flat("sin",self.SnoopWidth)
        parser.wid_op_flat("sdata",self.CamWidth)		        parser.wid_op_flat("sdata",self.CamWidth)
        parser.remove_sub_dict_flat("rd")			        parser.remove_sub_dict_flat("rd")
        self.get_all_key_value_pairs(parser.dict) 		        self.get_all_key_value_pairs(parser.dict) 

    def Create_dic_of_variable(self):				    def Create_dic_of_variable(self):

        self.variable_dict['CamWidth']=self.CamWidth		        self.variable_dict['CamWidth']=self.CamWidth
        self.variable_dict['SnoopWidth']=self.SnoopWidth	        self.variable_dict['SnoopWidth']=self.SnoopWidth
        self.variable_dict['CamDepth']= self.CamDepth 		        self.variable_dict['CamDepth']= self.CamDepth 
        self.variable_dict['EncodedDepth']=self.EncodedDepth	        self.variable_dict['EncodedDepth']=self.EncodedDepth
    								    
    def get_body(self):						    def get_body(self):
        dynamicgenerator=DynamicGenerator(self.variable_dict,	        dynamicgenerator=DynamicGenerator(self.variable_dict,
        self.body+=dynamicgenerator.parse_body()		        self.body+=dynamicgenerator.parse_body()
        self.body = self.body.replace("ENCODEDDEPTH - 1", str	        self.body = self.body.replace("ENCODEDDEPTH - 1", str
        self.body = self.body.replace("ENCODEDDEPTH", str(sel	        self.body = self.body.replace("ENCODEDDEPTH", str(sel
        self.body = self.body.replace("CAMWIDTH - 1", str(sel	        self.body = self.body.replace("CAMWIDTH - 1", str(sel
        self.body = self.body.replace("CAMWIDTH", str(self.Ca	        self.body = self.body.replace("CAMWIDTH", str(self.Ca
        self.body = self.body.replace("SNOOPWIDTH - 1", str(s	        self.body = self.body.replace("SNOOPWIDTH - 1", str(s
        self.body = self.body.replace("SNOOPWIDTH", str(self.	        self.body = self.body.replace("SNOOPWIDTH", str(self.
        dynamicgenerator.silentremove()				        dynamicgenerator.silentremove()


    def get_verilog(self):					    def get_verilog(self):
        modulecode=self.get_header()				        modulecode=self.get_header()
        self.get_body()						        self.get_body()
        modulecode=modulecode.replace("BODY",self.body)		        modulecode=modulecode.replace("BODY",self.body)
        return modulecode					        return modulecode

    def main(self):						    def main(self):
        self.write_to_file(self.get_verilog())			        self.write_to_file(self.get_verilog())
        return self.get_verilog()				        return self.get_verilog()


    def __init__(self, camdepth, camwidth, snoopwidth, path_o	    def __init__(self, camdepth, camwidth, snoopwidth, path_o
        self.CamDepth = camdepth				        self.CamDepth = camdepth
        self.EncodedDepth = ceil(log2(int(camdepth)))		        self.EncodedDepth = ceil(log2(int(camdepth)))
        self.SnoopWidth = snoopwidth				        self.SnoopWidth = snoopwidth
        self.CamWidth = camwidth				        self.CamWidth = camwidth
        self.name = "AH_" + self.__class__.__name__ + "_" + s	        self.name = "AH_" + self.__class__.__name__ + "_" + s
        super().__init__(self.name)				        super().__init__(self.name)
        self.body = ""						        self.body = ""
        self.variable_dict={}					        self.variable_dict={}
        self.Create_dic_of_variable()				        self.Create_dic_of_variable()
        self.add_ports_from_bus(path_of_yaml, bus_name)		        self.add_ports_from_bus(path_of_yaml, bus_name)
        self.cambody="""					        self.cambody="""
wire [ENCODEDDEPTH - 1:0] internal_wr_ptr;			wire [ENCODEDDEPTH - 1:0] internal_wr_ptr;
req  [ENCODEDDEPTH:0] wr_loc_counter;				req  [ENCODEDDEPTH:0] wr_loc_counter;
        /f_f/							        /f_f/
code = "\\n".join(["req  ["+str(CamWidth-1)+":0] cam_loc"+str	code = "\\n".join(["req  ["+str(CamWidth-1)+":0] cam_loc"+str
        /f_f/							        /f_f/
        always @ (posedge clk or negedge rst_an)		        always @ (posedge clk or negedge rst_an)
begin								begin
if (!rst_an) begin						if (!rst_an) begin
  wr_loc_counter <= ENCODEDDEPTH+1'd0; // log2(CAMDEPTH) +1	  wr_loc_counter <= ENCODEDDEPTH+1'd0; // log2(CAMDEPTH) +1
       end else begin						       end else begin
  wr_loc_counter <= wr_location_counter[ENCODEDDEPTH] ? wr_lo	  wr_loc_counter <= wr_location_counter[ENCODEDDEPTH] ? wr_lo
       end							       end
end								end
AH_srvfifo_ENCODEDDEPTH_CAMDEPTH u_wrloc_recirfifo (		AH_srvfifo_ENCODEDDEPTH_CAMDEPTH u_wrloc_recirfifo (
.wr_valid(freeup_loc)						.wr_valid(freeup_loc)
,.wr_ready(freedup_loc_valid)					,.wr_ready(freedup_loc_valid)
,.wr_data(freedup_loc_ready)					,.wr_data(freedup_loc_ready)

,.rd_valid(recir_loc_valid)					,.rd_valid(recir_loc_valid)
,.rd_ready(recir_loc_ready)					,.rd_ready(recir_loc_ready)
,.rd_data(recir_loc)						,.rd_data(recir_loc)
);								);

assign internal_wr_en = wr_valid & (recir_loc_valid | ~wr_loc	assign internal_wr_en = wr_valid & (recir_loc_valid | ~wr_loc
assign internal_wr_ptr = ~wr_location_counter[ENCODEDDEPTH + 	assign internal_wr_ptr = ~wr_location_counter[ENCODEDDEPTH + 

assign wr_ready	= ~wr_location_counter[ENCODEDDEPTH + 1] ? 1'	assign wr_ready	= ~wr_location_counter[ENCODEDDEPTH + 1] ? 1'

assign recir_loc_ready	= ~wr_location_counter[ENCODEDDEPTH +	assign recir_loc_ready	= ~wr_location_counter[ENCODEDDEPTH +

always @ (posedge clk or negedge rst_an)			always @ (posedge clk or negedge rst_an)
if (!rst_an) begin						if (!rst_an) begin

/f_f/								/f_f/
code = "\\n\t"+"\\n\t".join(["cam_loc"+str(i)+" <= "+str(CamW	code = "\\n\t"+"\\n\t".join(["cam_loc"+str(i)+" <= "+str(CamW
/f_f/								/f_f/

end else begin							end else begin
/f_f/								/f_f/
code = "\\n\t"+"\\n\t".join(["cam_loc"+str(i)+" <= (internal_	code = "\\n\t"+"\\n\t".join(["cam_loc"+str(i)+" <= (internal_
"'d"+str(i)+") ) ? wr_data : cam_loc"+str(i)+";" for i in ran	"'d"+str(i)+") ) ? wr_data : cam_loc"+str(i)+";" for i in ran
/f_f/								/f_f/

end								end

wire assign snoop_match = freedup_loc_ready) & (		wire assign snoop_match = freedup_loc_ready) & (

/f_f/								/f_f/
code = " |\\n ".join(["(snoop_in == cam_loc"+str(i)+"["+str(S	code = " |\\n ".join(["(snoop_in == cam_loc"+str(i)+"["+str(S
/f_f/								/f_f/
);								);

wire assign snoop_data = CAMWIDTH'd0 |				wire assign snoop_data = CAMWIDTH'd0 |
/f_f/								/f_f/
code = " |\\n ".join(["( (snoop_in == cam_loc"+str(i)+"["+str	code = " |\\n ".join(["( (snoop_in == cam_loc"+str(i)+"["+str
/f_f/								/f_f/
;								;

wire assign freeup_loc = ENCODEDDEPTH'd0 |			wire assign freeup_loc = ENCODEDDEPTH'd0 |
/f_f/								/f_f/
code = " |\\n ".join(["( (snoop_in == cam_loc"+str(i)+"["+str	code = " |\\n ".join(["( (snoop_in == cam_loc"+str(i)+"["+str
/f_f/								/f_f/
;								;

wire assign freedup_loc_valid = snoop_match;			wire assign freedup_loc_valid = snoop_match;
"""								"""
cam=CAM(int(sys.argv[1]), int(sys.argv[2]), int(sys.argv[3]),	cam=CAM(int(sys.argv[1]), int(sys.argv[2]), int(sys.argv[3]),
cam.main()							cam.main()
