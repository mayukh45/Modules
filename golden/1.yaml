decoder:
    dec_err:
        cname: dec_err
        direction: output
        heiarchy: decoder_dec_err
        name: dec_err
        type: fluid
        width: 1
    decoded_binary:
        cname: decoded_binary
        direction: output
        heiarchy: decoder_decoded_binary
        name: decoded_binary
        type: fluid
        width: 3
    ingress_pkt_field:
        cname: ingress_pkt_field
        direction: input
        heiarchy: decoder_ingress_pkt_field
        name: ingress_pkt_field
        type: fluid
        width: 32
